server/src/mds3.gdc.js

```diff
@@ -1282,15 +1284,110 @@ returns:
+
 	if (q.isHierCluster) {
-		/*
-		running gene exp clustering
-		must only restrict to cases with exp data
-		but not by mutated cases anymore, thus geneTwLst should not be used (and not supplied)
+		// running gene exp clustering, must only restrict to cases with exp data, but not by mutated cases anymore, thus geneTwLst should not be used (and not supplied)
+		;[byTermId, samples] = await querySamplesTwlstForGeneexpclustering(q, dictTwLst, ds)
+	} else {
+		// not in gene exp clustering mode
+		;[byTermId, samples] = await querySamplesTwlstNotForGeneexpclustering(q, dictTwLst, ds, geneTwLst)
+	}
+
+	// step 3: querying survival data
+
+	if (survivalTwLst.length) {
+		await querySamplesSurvival(q, survivalTwLst, ds, samples)
+	}
+
+	return { byTermId, samples }
+}
+
+async function querySamplesSurvival(q, survivalTwLst, ds, samples) {
+	// build survival api filter
+	// XXX FIXME may not be sufficiently limiting cases by filter
+	const filter = { op: 'and', content: [] }
+	if (q.filter0) {
+		filter.content.push(q.filter0)
+	}
+
+	if (q.set_id) {
+		if (typeof q.set_id != 'string') throw '.set_id value not string'
+		filter.content.push({
+			op: 'in',
+			content: {
+				field: 'cases.case_id',
+				value: [q.set_id]
+			}
+		})
+	}
+	if (q.filterObj) {
+		const g = filter2GDCfilter(q.filterObj)
+		if (g) filter.content.push(g)
+	}
+	addTid2value_to_filter(q.tid2value, filter.content, ds)
+
+	// the survival term itself is not used in api query, since there's just one type of survival data from gdc and no need to distinguish
+	const { host, headers } = ds.getHostHeaders(q)
+	const re = await ky
+		.post('https://portal.gdc.cancer.gov/auth/api/v0/analysis/survival', {
+			timeout: false,
+			headers,
+			json: { filters: [filter] }
+		})
+		.json()
+	if (!Array.isArray(re.results?.[0].donors)) throw 're.results[0].donors[] not array'
+	for (const d of re.results[0].donors) {
+		/* each d:
+		{
+          "time": 2.0, // time to event in #days
+          "censored": true,
+          "survivalEstimate": 1, // percentage in 0-1 range, may be for km plot. no use here
+          "id": "e117fcf4-2c01-4b4d-8faf-5de7d0e839f1", // case uuid
+          "submitter_id": "TCGA-HT-A5R9",
+          "project_id": "TCGA-LGG"
+        },
 		*/
-		return await querySamplesTwlst4hierCluster(q, twLst, ds)
+		if (!d.id) throw 'd.id (case uuid) missing'
+		if (typeof d.censored != 'boolean') throw 'd.censored is not boolean'
+		if (!Number.isFinite(d.time)) throw 'd.time not number'
+		let s = samples.find(i => i.sample_id == d.id) // find existing sample
+		if (!s) {
+			// may create new and insert to samples[]
+			if (q.isHierCluster && !ds.__gdc.casesWithExpData.has(d.id)) {
+				// case does not have gene exp data; only cases with exp data are allowed
+				continue
+			}
+			s = { sample_id: d.id }
+			samples.push(s)
+		}
+		// assign survival data for this sample
+		s[survivalTwLst[0].term.id] = {
+			key: d.censored ? 1 : 0,
```

Comment:
@congyu-lu please see if the key value 1/0 is assigned correctly

Reply:
I think censored generally has a code of 0 and event-occurred has a code of 1